ffi/firewood.h

```diff
@@ -20,24 +20,78 @@
 typedef struct DatabaseHandle DatabaseHandle;
 
 /**
- * A value returned by the FFI.
- *
- * This is used in several different ways, including:
- * * An C-style string.
- * * An ID for a proposal.
- * * A byte slice containing data.
- *
- * For more details on how the data may be stored, refer to the function signature
- * that returned it or the `From` implementations.
- *
- * The data stored in this struct (if `data` is not null) must be manually freed
- * by the caller using `fwd_free_value`.
+ * A root database hash key, used in FFI functions that provide or return
+ * a root hash. This type requires no allocation and can be copied freely
+ * and droppped without any additional overhead.
+ *
+ * This is useful because it is the same size as 4 words which is equivalent
+ * to 2 heap-allocated slices (pointer + length each), or 1.5 vectors (which
+ * uses an extra word for allocation capacity) and it can be passed around
+ * without needing to allocate or deallocate memory.
+ */
+typedef struct HashKey {
+  uint8_t _0[32];
+} HashKey;
+
+/**
+ * A Rust-owned vector of bytes that can be passed to C code.
  *
+ * C callers must free this memory using the respective FFI function for the
+ * concrete type (but not using the `free` function from the C standard library).
  */
-typedef struct Value {
+typedef struct OwnedSlice_u8 {
+  uint8_t *ptr;
   size_t len;
-  uint8_t *data;
-} Value;
+} OwnedSlice_u8;
+
+/**
+ * A type alias for a rust-owned byte slice.
+ */
+typedef struct OwnedSlice_u8 OwnedBytes;
+
+/**
+ * A result type returned from FFI functions return the database root hash. This
+ * may or may not be after a mutation.
+ */
+typedef enum HashResult_Tag {
+  /**
+   * The caller provided a null pointer to a database handle.
+   */
+  HashResult_NullHandlePointer,
```

Comment:
This seems pretty uncommon. Why can't we just make this another HashResult_Err?

Reply:
Ideally all errors would eventually propagate as an enum instead of allocating a string, but I started with the NPE because it's the most common variant. The generated asm code surrounding handling the NPE also be more efficient as it will go from zero input to zero output.